sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -1529,9 +1529,11 @@ pi_result piPlatformsGet(pi_uint32 NumEntries, pi_platform *Platforms,
     PrintPiTrace = true;
   }
 
-  if (ZeDebug & ZE_DEBUG_CALL_COUNT) {
-    ZeCallCount = new std::map<const char *, int>;
-  }
+  static std::once_flag ZeCallCountInitialized;
+  std::call_once(ZeCallCountInitialized, []() {
```

Comment:
This call can throw, so need to catch

Reply:
I don't think std::call_once() can throw.
Do you mean new in lambda can throw?

Reply:
If the function itself fails, it throws a system_error exception. And also the called function's exceptions are re-thrown.

Reply:
The lambda includes only `new` operation.
Are you saying I have to try-catch the `new` operator?
Once we catch the exception, what you do want to do to recover?
Not all exceptions are recoverable, and usually exception from `new` should just crash because we can't do anything to overcome this exception.

Reply:
Use PI_OUT_OF_HOST_MEMORY for out of memory, PI_ERROR_UNKNOWN for everything else. See https://github.com/intel/llvm/blob/e66216673f599d85c056bc3a9dd7ddfd5544ff3d/sycl/plugins/level_zero/pi_level_zero.cpp#L1906

Reply:
How about this?
```
  static std::once_flag ZeCallCountInitialized;
  std::call_once(ZeCallCountInitialized, []() {
    if (ZeDebug & ZE_DEBUG_CALL_COUNT) {
      try {
        ZeCallCount = new std::map<const char *, int>;
      } catch (const std::bad_alloc &) {
        return PI_OUT_OF_HOST_MEMORY;
      }
  });
```

Reply:
this is still not catching the system_error exception possible from the "call_once" itself

Reply:
Thanks. How about this?
```
  static std::once_flag ZeCallCountInitialized;
  try {
    std::call_once(ZeCallCountInitialized, []() {
      if (ZeDebug & ZE_DEBUG_CALL_COUNT) {
        ZeCallCount = new std::map<const char *, int>;
      }
    });
  } catch (std::system_error &) {
    return PI_OUT_OF_HOST_MEMORY;
  }
```

Reply:
This one doesn't catch the "std::bad_alloc" possible from the new. Just catch all

Reply:
Thanks. I changed to catch all possible exceptions.
```
  static std::once_flag ZeCallCountInitialized;
  try {
    std::call_once(ZeCallCountInitialized, []() {
      if (ZeDebug & ZE_DEBUG_CALL_COUNT) {
        ZeCallCount = new std::map<const char *, int>;
      }
    });
  } catch (...) {
    return PI_OUT_OF_HOST_MEMORY;
  }
```